### ISSUE TITLE
Execute a cell decorator

### ISSUE THREAD

**Issue by MSeal** [nteract/testbook#5]
Add the most basic "execute indicated cell" before unittest runs decorator. This will be used to prep a test case for evaluation.

**Comment by rohitsanj** [comment]
Hey, I'm looking into this. 

So are we looking at some functionality like..
```python
@execute_cell(3)
def test_foo():
  pass
```
If yes, 
How and where do we define which notebook file is being executed?

**Comment by MSeal** [comment]
Yes something like this would be good.

`testbook.execute_cell` is correct, but you can assume people may do `from testbook import execute_cell` to get back to the function name alone.

I see two ways we can support which notebook to be executing. First is to require a separate decorator first (e.g. `@testbook.notebook(path_to_notebook)` which adds the `notebook` keyword argument to the function call) or as as an optional argument to `execute_cell(path_to_notebook, 3)`. I imagine we'll want to support both patterns, but we can just use one for now to make it easier. Either one should add `notebook` as an argument to the `test_foo(notebook)` call so the in-memory object can be accessed / tested. At first this can just be the nbformat / nbclient response but we likely will want to make a wrapper objects with test assertion calls available.

**Comment by rohitsanj** [comment]
> I imagine we'll want to support both patterns, but we can just use one for now to make it easier.
Alright, I'll go ahead with adding an optional argument for now. 

> At first this can just be the nbformat / nbclient response but we likely will want to make a wrapper objects with test assertion calls available.
I'm not sure what you meant by this? 😅

**Comment by MSeal** [comment]
>> At first this can just be the nbformat / nbclient response but we likely will want to make a wrapper objects with test assertion calls available.

> I'm not sure what you meant by this? sweat_smile

So when you load (or execute) a notebook you get this nbformat object which is basically a dict holding the v4 notebook schema. This contains things like tags, cells, outputs, metadata, etc. In our testbook case we want to expose the values therein so people can make assertions against the fields. But we likely will want to add additional functions like `notebook.assert_output_text('I printed this message')` for convienence. This means we'll want to make a new wrapping object which mimics or inherits from nbformat's document kind of like https://github.com/nteract/scrapbook/blob/master/scrapbook/models.py#L40

**Comment by rohitsanj** [comment]
So I tried to use `NotebookClient.execute_cell` as follows:

```python
# assume that the following is inside a decorator
with open(notebook_filename) as f:
    nb = nbformat.read(f, as_version=4)

client = NotebookClient(nb)
client.execute_cell(nb['cells'][cell_index], cell_index)

@execute_cell('../testbook_example.ipynb', 2)
def foo():
    pass

foo()
```
It throws an `AttributeError`
```
File "/Users/rohitsanjay/miniconda3/envs/testbook/lib/python3.7/site-packages/nbclient/client.py", line 528, in execute_cell
    parent_msg_id = self.kc.execute(
AttributeError: 'NoneType' object has no attribute 'execute'
```

I looked into the code in `client.py` and found that [`setup_kernel`](https://github.com/jupyter/nbclient/blob/96148fdea41cb78cd1d8ad635019348824e85f98/nbclient/client.py#L393) is not being used in [`execute_cell`](https://github.com/jupyter/nbclient/blob/96148fdea41cb78cd1d8ad635019348824e85f98/nbclient/client.py#L553) and hence in [`async_execute_cell`](https://github.com/jupyter/nbclient/blob/96148fdea41cb78cd1d8ad635019348824e85f98/nbclient/client.py#L592). 

I'm not sure if this a bug or not. @MSeal

**Comment by rohitsanj** [comment]
> > > At first this can just be the nbformat / nbclient response but we likely will want to make a wrapper objects with test assertion calls available.
> 
> > I'm not sure what you meant by this? sweat_smile
> 
> So when you load (or execute) a notebook you get this nbformat object which is basically a dict holding the v4 notebook schema. This contains things like tags, cells, outputs, metadata, etc. In our testbook case we want to expose the values therein so people can make assertions against the fields. But we likely will want to add additional functions like `notebook.assert_output_text('I printed this message')` for convienence. This means we'll want to make a new wrapping object which mimics or inherits from nbformat's document kind of like https://github.com/nteract/scrapbook/blob/master/scrapbook/models.py#L40

Thanks for the explanation. It makes perfect sense now 😄

**Comment by MSeal** [comment]
> I looked into the code in client.py and found that setup_kernel is not being used in execute_cell and hence in async_execute_cell.
>
> I'm not sure if this a bug or not. @MSeal

We're refactoring that code a bit -- make sure to use the released 0.1 code for now -- we'll be keeping that contract as best we can before 0.2 releases.

For now I would look at how papermill extends it to prep/wrap nbclient: https://github.com/nteract/papermill/blob/master/papermill/clientwrap.py#L30-L70. We can also update nbclient to have better extension interfaces in the future as we progress.

**Comment by rohitsanj** [comment]
I'll create a `TestbookNotebookClient` that extends `NotebookClient` then.
Thanks.
